### DISCUSSION TITLE
Web XR behavior in Meta Quest 2

### DISCUSSION THREAD

**Discussion by IRobot1** [nartc/angular-three#132] [Q&A]
First, this problem doesn't happen using Windows Mixed Reality in Edge browser.  I can only repeat using Meta Quest 2 after Entering VR from its browser

My app page looks like this
```html
  <ngt-canvas webxr (created)="created($event)" [camera]="{ fov: 55, position: [0, 2, 4]}">
    <router-outlet></router-outlet>
  </ngt-canvas>
```
Pages loaded via router changing the contents of the canvas.  Here's a basic example of new content being added to the canvas
```html
<ngt-color attach="background" color="darkgray"></ngt-color>

<ngt-directional-light [position]="[10, 10, 10]" castShadow></ngt-directional-light>

<ngt-mesh>
    <ngt-mesh-standard-material></ngt-mesh-standard-material>
    <ngt-sphere-geometry [args]="[0.2]"></ngt-sphere-geometry>
</ngt-mesh>
```
When navigate between router inside the Quest browser, the pages render correctly.  However, when I Enter VR and navigate to a new content via a route, the canvas is blank

Using the example above, nothing is render; not the light or the mesh.
If I render using THREE objects, they're added to the scene properly.
I've tried navigating without using router, problem still happens.
I even tried forcing change detection on every tick

Its like the NG3 change detection completely stops or gets stuck rendering all changes outside of Angular zone.

I've installed the Oculus development tools and remotely monitored the browser console.  No errors or warnings.

I've spent days trying to debug this problem and finally out of things to try.

Any suggestions?  I'll try putting the most basic repeat case together and sharing.

If you have a Meta Quest headset, you can see the problem in this [DEMO](https://ng3vr.z9.web.core.windows.net/)

**Comment by nartc** [comment]
Hi there @IRobot1 , sorry in advance. I don't have a VR set to test this out as well as I don't have enough time nowadays to work on AngularThree much.

**Reply by IRobot1** [reply]
I think I've got to the bottom of it.  

[VRButton](https://github.com/mrdoob/three.js/blob/1bf474b0317573481619149f60a113457c9d38c2/examples/jsm/webxr/VRButton.js#L71) requests support for WebXR [layers](https://developer.mozilla.org/en-US/docs/Web/API/XRRenderState/layers).

However, threejs WebXRManager [does not properly handle](https://github.com/mrdoob/three.js/blob/1bf474b0317573481619149f60a113457c9d38c2/src/renderers/webxr/WebXRManager.js#L243) WebXR layers

On Windows Mixed Reality, layers are not supported by Edge browser. The following warning appears in the browser console `Unrecognized feature requested: layers`.  Immersive Web [Layer Samples](https://immersive-web.github.io/webxr-samples/layers-samples/) fail with required feature not support error.  I've confirmed in the renderState than layers array is always undefined.

However, on Meta Quest browser, layers are supported and being used.  I've confirmed renderState layers array is 1.  Unfortunately, threejs does not call WebXR layers updateRenderState on these layers.

I have two options
1) wait for threejs to support WebXR layers
2) roll my own VR button that doesn't ask for layers and hope the Meta Quest browser honors it

**Reply by IRobot1** [reply]
I noticed in NGT, that [gl.xr animation loop is overriden](https://github.com/nartc/angular-three/blob/8feef75f7633ff35f079ad84dae678a0749d1a25/libs/core/src/lib/stores/store.ts#L364).

Doesn't this still need to call the [original animation loop in threejs](https://github.com/mrdoob/three.js/blob/1bf474b0317573481619149f60a113457c9d38c2/src/renderers/webxr/WebXRManager.js#L717)?

**Reply by IRobot1** [reply]
Never mind.  Looks like its a hook / callback method.  So NGT gets called after every animation frame.

**Reply by IRobot1** [reply]
Can you think of any conditions under which beforeRender stops getting called?  
Prior to entering VR, objects with beforeRender callbacks are getting called.  
On the Quest, after Enter VR, beforeRender stops getting called.

**Reply by IRobot1** [reply]
Found the root cause;  [requestAnimationFrame](https://github.com/nartc/angular-three/blob/8feef75f7633ff35f079ad84dae678a0749d1a25/libs/core/src/lib/abstracts/object.ts#L447). When in VR on the Quest, this method never gets called and makes sense why nothing gets added to the scene.

From my understanding, when in xr.enabled = true, [XRSession.requestAnimationFrame](https://developer.mozilla.org/en-US/docs/Web/API/XRSession/requestAnimationFrame) should be called.

**Reply by IRobot1** [reply]
Yes, that was it.  I'll submit a bug and pull request.
